internal/state/documents.go

```diff
@@ -0,0 +1,166 @@
+package state
+
+import (
+	"log"
+	"time"
+
+	"github.com/hashicorp/go-memdb"
+	"github.com/hashicorp/terraform-ls/internal/document"
+	"github.com/hashicorp/terraform-ls/internal/source"
+)
+
+type DocumentStore struct {
+	db        *memdb.MemDB
+	tableName string
+	logger    *log.Logger
+
+	// TimeProvider provides current time (for mocking time.Now in tests)
+	TimeProvider func() time.Time
+}
+
+func (s *DocumentStore) OpenDocument(dh document.Handle, langId string, version int, text []byte) error {
+	txn := s.db.Txn(true)
+	defer txn.Abort()
+
+	// TODO: Introduce Exists method to Txn?
+	obj, err := txn.First(s.tableName, "id", dh.Dir, dh.Filename)
+	if err != nil {
+		return err
+	}
+	if obj != nil {
+		return &AlreadyExistsError{
+			Idx: dh.FullURI(),
+		}
+	}
+
+	doc := &document.Document{
+		Dir:        dh.Dir,
+		Filename:   dh.Filename,
+		ModTime:    s.TimeProvider(),
+		LanguageID: langId,
+		Version:    version,
+		Text:       text,
+		Lines:      source.MakeSourceLines(dh.Filename, text),
+	}
+
+	err = txn.Insert(s.tableName, doc)
+	if err != nil {
+		return err
+	}
+
+	txn.Commit()
+	return nil
+}
+
+func (s *DocumentStore) UpdateDocument(dh document.Handle, newText []byte, newVersion int) error {
+	txn := s.db.Txn(true)
+	defer txn.Abort()
+
+	doc, err := copyDocument(txn, dh)
+	if err != nil {
+		return err
+	}
+
+	doc.Text = newText
+	doc.Lines = source.MakeSourceLines(dh.Filename, newText)
+	doc.Version = newVersion
+	doc.ModTime = s.TimeProvider()
+
+	err = txn.Insert(s.tableName, doc)
+	if err != nil {
+		return err
+	}
+
+	txn.Commit()
+	return nil
+}
+
+func copyDocument(txn *memdb.Txn, dh document.Handle) (*document.Document, error) {
+	doc, err := getDocument(txn, dh)
+	if err != nil {
+		return nil, err
+	}
+
+	return doc.Copy(), nil
+}
+
+func (s *DocumentStore) GetDocument(dh document.Handle) (*document.Document, error) {
+	txn := s.db.Txn(false)
+	return getDocument(txn, dh)
+}
+
+func getDocument(txn *memdb.Txn, dh document.Handle) (*document.Document, error) {
+	obj, err := txn.First(documentsTableName, "id", dh.Dir, dh.Filename)
+	if err != nil {
+		return nil, err
+	}
+	if obj == nil {
+		return nil, &document.DocumentNotFound{
+			URI: dh.FullURI(),
+		}
+	}
+	return obj.(*document.Document), nil
+}
+
+func (s *DocumentStore) CloseDocument(dh document.Handle) error {
+	txn := s.db.Txn(true)
+	defer txn.Abort()
+
+	obj, err := txn.First(s.tableName, "id", dh.Dir, dh.Filename)
```

Comment:
Would it be possible to use `getDocument` here? The operation looks identical.

Reply:
It is _almost_ identical. The main use case for `getDocument` is to get the contents, not check existence (which is what we need to do in `CloseDocument`).

It's probably negligible benefit in the context of a single document but it might make some difference at scale (e.g. should the user close many documents in quick succession) as we avoid the type casting in `CloseDocument` from `interface{}` to `*document.Document`.